scio-test/src/main/scala/com/spotify/scio/testing/TransformFinder.scala

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright 2023 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.spotify.scio.testing
+
+import org.apache.beam.sdk.Pipeline.PipelineVisitor
+import org.apache.beam.sdk.Pipeline.PipelineVisitor.CompositeBehavior
+import org.apache.beam.sdk.runners.{PTransformMatcher, TransformHierarchy}
+import org.apache.beam.sdk.transforms.PTransform
+
+import scala.collection.mutable
+
+class TransformFinder(matcher: PTransformMatcher) extends PipelineVisitor.Defaults {
```

Comment:
Some scaladoc wouldn't hurt. Not quite sure what `freed` means in this context... non-matching?